lib/iris/analysis/_grid_angles.py

```diff
@@ -0,0 +1,390 @@
+        x_coord, y_coord = x.copy(), y.copy()
+        x_coord.convert_units('degrees')
+        y_coord.convert_units('degrees')
+        if x_coord.ndim != 2 or y_coord.ndim != 2:
+            msg = ('Coordinate inputs must have 2-dimensional shape. ',
+                   'Got x-shape of {} and y-shape of {}.')
+            raise ValueError(msg.format(x_coord.shape, y_coord.shape))
+        if x_coord.shape != y_coord.shape:
+            msg = ('Coordinate inputs must have same shape. ',
+                   'Got x-shape of {} and y-shape of {}.')
+            raise ValueError(msg.format(x_coord.shape, y_coord.shape))
+# NOTE: would like to check that dims are in correct order, but can't do that
+# if there is no cube.
+# TODO: **document** -- another input format requirement
+#        x_dims, y_dims = (cube.coord_dims(co) for co in (x_coord, y_coord))
+#        if x_dims != (0, 1) or y_dims != (0, 1):
+#            msg = ('Coordinate inputs must map to cube dimensions (0, 1). ',
+#                   'Got x-dims of {} and y-dims of {}.')
+#            raise ValueError(msg.format(x_dims, y_dims))
+        if x_coord.has_bounds() and y_coord.has_bounds():
+            x, y = x_coord.bounds, y_coord.bounds
+        else:
+            x, y = x_coord.points, y_coord.points
+
+    elif hasattr(x, 'bounds') or hasattr(y, 'bounds'):
+        # One was a Coord, and the other not ?
+        is_and_not = ('x', 'y')
+        if hasattr(y, 'bounds'):
+            is_and_not = reversed(is_and_not)
+        msg = 'Input {!r} is a Coordinate, but {!r} is not.'
+        raise ValueError(*is_and_not)
+
+    # Now have either 2 points arrays or 2 bounds arrays.
+    # Construct (lhs, mid, rhs) where these represent 3 adjacent points with
+    # increasing longitudes.
+    if x.ndim == 2:
+        # PROBLEM: we can't use this if data is not full-longitudes,
+        # i.e. rhs of array must connect to lhs (aka 'circular' coordinate).
+        # But we have no means of checking that ?
+
+        # Use previous + subsequent points along longitude-axis as references.
+        # NOTE: we also have no way to check that dim #2 really is the 'X' dim.
+        mid = np.array([x, y])
+        lhs = np.roll(mid, 1, 2)
+        rhs = np.roll(mid, -1, 2)
+        if not x_coord:
+            # Create coords for result cube : with no bounds.
+            y_coord = iris.coords.AuxCoord(x, standard_name='latitude',
+                                           units='degrees')
+            x_coord = iris.coords.AuxCoord(y, standard_name='longitude',
+                                           units='degrees')
+    else:
+        # Get lhs and rhs locations by averaging top+bottom each side.
+        # NOTE: so with bounds, we *don't* need full circular longitudes.
+        xyz = _3d_xyz_from_latlon(x, y)
+        angle_boundpoints_vals = {'mid-lhs, mid-rhs': '03_to_12',
+                                  'lower-left, lower-right': '0_to_1'}
```

Comment:
Kinda confusing names `03_to_12` and `0_to_1`. I initially thought it may have something to do with a clock.
I understand them now, but I will try to think up something informative